packages/fuels-types/src/param_types.rs

```diff
@@ -67,42 +68,75 @@ impl ParamType {
+    pub fn calculate_num_of_elements(&self, bytes: &[u8]) -> Result<usize> {
+        let memory_size = match self {
+            ParamType::Bytes | ParamType::Vector(..) => Ok(self.heap_inner_element_size().unwrap()),
+            ParamType::RawSlice => Ok(ParamType::U64.compute_encoding_width() * WORD_SIZE),
+            _ => Err(error!(
+                InvalidData,
+                "Method `calculate_num_of_elements` called on an incompatible type: {self:?}"
+            )),
+        }
+        .unwrap();
+        if bytes.len() % memory_size != 0 {
+            return Err(error!(
+                InvalidData,
+                "The bytes provided do not correspond to a {:?} got: {:?}", self, bytes
+            ));
+        }
+        Ok(bytes.len() / memory_size)
+    }
+
+    pub fn contains_nested_heap_types(&self) -> bool {
         match &self {
-            ParamType::Vector(param_type) => param_type.uses_vectors(),
-            _ => self.uses_vectors(),
+            ParamType::Vector(param_type) => param_type.uses_heap_types(),
+            ParamType::Bytes => false,
+            _ => self.uses_heap_types(),
         }
     }
 
-    fn uses_vectors(&self) -> bool {
+    fn uses_heap_types(&self) -> bool {
         match &self {
             ParamType::Vector(..) => true,
-            ParamType::Array(param_type, ..) => param_type.uses_vectors(),
-            ParamType::Tuple(param_types, ..) => Self::any_nested_vectors(param_types),
+            ParamType::Bytes => true,
+            ParamType::Array(param_type, ..) => param_type.uses_heap_types(),
+            ParamType::Tuple(param_types, ..) => Self::any_nested_heap_types(param_types),
             ParamType::Enum {
                 generics, variants, ..
             } => {
                 let variants_types = variants.param_types();
-                Self::any_nested_vectors(chain!(generics, &variants_types))
+                Self::any_nested_heap_types(chain!(generics, &variants_types))
             }
             ParamType::Struct {
                 fields, generics, ..
             } => {
                 let fields = fields.iter().map(|(_, param_type)| param_type);
-                Self::any_nested_vectors(chain!(fields, generics))
+                Self::any_nested_heap_types(chain!(fields, generics))
             }
             _ => false,
         }
     }
 
-    fn any_nested_vectors<'a>(param_types: impl IntoIterator<Item = &'a ParamType>) -> bool {
+    fn any_nested_heap_types<'a>(param_types: impl IntoIterator<Item = &'a ParamType>) -> bool {
         param_types
             .into_iter()
-            .any(|param_type| param_type.uses_vectors())
+            .any(|param_type| param_type.uses_heap_types())
     }
 
-    pub fn is_vector(&self) -> bool {
-        matches!(self, Self::Vector(..))
+    pub fn is_vm_heap_type(&self) -> bool {
+        matches!(self, ParamType::Vector(..)) || matches!(self, ParamType::Bytes)
```

Comment:
Can be replaced by one `matches!`:
```suggestion
        matches!(self, ParamType::Vector(..) | ParamType::Bytes)
```